### PR TITLE
perf: optimisations for ont - barcode+adapter-trimming, primer-trimming (porechop) and read correction (Canu)

### PR DESCRIPTION
**Pull request by simakro**
# Description

perf: Optimizations for barcode+adapter-trimming, primer-trimming (porechop) and read correction (Canu)

## Checklist

- [x] I've formatted the PR title in accordance with the [structure of the conventional commit messages](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I've updated the code style using [`pre-commit`](https://ikim-essen.github.io/uncovar/dev-guide/contributing/#pre-commit) if needed.
- [x] I've updated or supplemented the documentation as needed.
- [x] I've read the [`CODE_OF_CONDUCT.md`] document.
- [x] I've read the [`CONTRIBUTING.md`] guide.

[`CODE_OF_CONDUCT.md`]: https://github.com/IKIM-Essen/uncovar/blob/master/CODE_OF_CONDUCT.md
[`CONTRIBUTING.md`]: https://github.com/IKIM-Essen/uncovar/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by thomasbtf** [comment]
Hey Simon, thanks for the PR and the performance improvements! Could you please:
- formatted the PR title in accordance with the [structure of the conventional commit messages](https://www.conventionalcommits.org/en/v1.0.0/) (see the comments in PRs body for examples) and
- updated the code style using [pre-commit](https://ikim-essen.github.io/uncovar/dev-guide/contributing/#pre-commit).

Thanks!

File: workflow/rules/long_read.smk [hunk 1]
```diff
@@ -104,20 +111,35 @@ rule canu_correct:
         "logs/{date}/canu/assemble/{sample}.log",
     params:
         outdir=get_output_dir,
+        # concurrency=lambda w, threads: int(threads / 4),
```
Comment (thomasbtf) [inline workflow/rules/long_read.smk:114]:
```suggestion
```

File: workflow/rules/long_read.smk [hunk 2]
```diff
@@ -104,20 +111,35 @@ rule canu_correct:
...
+        concurrency=lambda w, threads: get_canu_concurrency(threads),
         min_length=config["quality-criteria"]["ont"]["min-length-reads"],
         for_testing=lambda w, threads: get_if_testing(
-            f"corThreads={threads} redThreads={threads} redMemory=6 oeaMemory=6"
+            f"corThreads={threads} redMemory=6 oeaMemory=6"
         ),
+        # redThreads={threads}
```
Comment (thomasbtf) [inline workflow/rules/long_read.smk:120]:
```suggestion
```

**Review by thomasbtf — APPROVED** [review]
LGTM!

Edit: I just had a closer look at `get_canu_concurrency`. See below.

File: workflow/rules/common.smk [hunk 3]
```diff
@@ -1046,6 +1046,14 @@ def get_output_dir(wildcards, output):
     return os.path.dirname(output[0])
 
 
+def get_canu_concurrency(threads):
+    if threads > 3:
+        lambda w, threads: int(threads / 4)
```
Comment (thomasbtf) [inline workflow/rules/common.smk:1051]:
```suggestion
        threads = int(threads / 4)
```

**Review by thomasbtf — CHANGES_REQUESTED** [review]

File: workflow/rules/common.smk (hunk 3 above)
Comment (simakro) [reply workflow/rules/common.smk:1051]: True, the lambda is of course not needed here. I took it like that from the rule.
Actually there is an even more concise way by just returning int(threads / 4).
Ill do it this way.

**Review by thomasbtf — APPROVED** [review]

File: workflow/rules/common.smk (hunk 3 above)
Comment (thomasbtf) [reply workflow/rules/common.smk]: Sound good to me 👍

**Pull request merged**